### PR TITLE
Correct SNR weighted loss in v-prediction case by only adding 1 to SNR on the denominator

### LINKED ISSUE #5654: Min-SNR weight for v-prediction in examples
In #5238, the [Min-SNR](https://openaccess.thecvf.com/content/ICCV2023/papers/Hang_Efficient_Diffusion_Training_via_Min-SNR_Weighting_Strategy_ICCV_2023_paper.pdf) weight for v-prediction is implemented as 
$$w_t = \frac {min(\text{SNR}+1, \gamma)} {\text{SNR}+1}$$ 
but it should be
$$w_t = \frac {min(\text{SNR}, \gamma)} {\text{SNR}+1}$$.
See https://github.com/huggingface/diffusers/blob/main/examples/text_to_image/train_text_to_image.py#L931-L937.

### PR DESCRIPTION
**Pull request by thuliu-yt16**
# What does this PR do?

Fixes #5654 (issue)

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case. (#5654)

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@bghira

### CODE REVIEW

**Comment by bghira** [comment]
should also be able to fix sample prediction_type (x-pred)

File: examples/controlnet/train_controlnet_flax.py [hunk 1]
```diff
@@ -907,10 +907,15 @@ def compute_loss(params, minibatch, sample_rng):
...
-                    snr = snr + 1
-                snr_loss_weights = jnp.where(snr < args.snr_gamma, snr, jnp.ones_like(snr) * args.snr_gamma) / snr
+                if noise_scheduler.config.prediction_type == "epsilon":
+                    snr_loss_weights = jnp.where(snr < args.snr_gamma, snr, jnp.ones_like(snr) * args.snr_gamma) / snr
+                elif noise_scheduler.config.prediction_type == "v_prediction":
+                    snr_loss_weights = jnp.where(snr < args.snr_gamma, snr, jnp.ones_like(snr) * args.snr_gamma) / (
+                        snr + 1
+                    )
+                else:
+                    raise ValueError(f"Unknown prediction type {noise_scheduler.config.prediction_type}")
```
Comment (bghira) [inline examples/controlnet/train_controlnet_flax.py:917]: technically it supported prediction_type = "sample" before, since the only special case was v-prediction.

should be able to put that back, as the prediction type check is earlier.

**Review by bghira — CHANGES_REQUESTED** [review]

File: examples/controlnet/train_controlnet_flax.py [hunk 2]
```diff
@@ -907,10 +907,15 @@ def compute_loss(params, minibatch, sample_rng):
 
             if args.snr_gamma is not None:
                 snr = jnp.array(compute_snr(timesteps))
-                if noise_scheduler.config.prediction_type == "v_prediction":
-                    # Velocity objective requires that we add one to SNR values before we divide by them.
-                    snr = snr + 1
-                snr_loss_weights = jnp.where(snr < args.snr_gamma, snr, jnp.ones_like(snr) * args.snr_gamma) / snr
+                if noise_scheduler.config.prediction_type == "epsilon":
+                    snr_loss_weights = jnp.where(snr < args.snr_gamma, snr, jnp.ones_like(snr) * args.snr_gamma) / snr
```
Comment (sayakpaul) [inline examples/controlnet/train_controlnet_flax.py:911]: Why do we have to compute `snr_loss_weights` two times?

File: examples/kandinsky2_2/text_to_image/train_text_to_image_decoder.py [hunk 3]
```diff
@@ -781,12 +781,16 @@ def collate_fn(examples):
...
-                        torch.stack([snr, args.snr_gamma * torch.ones_like(timesteps)], dim=1).min(dim=1)[0] / snr
-                    )
+                    if noise_scheduler.config.prediction_type == "epsilon":
+                        mse_loss_weights = (
+                            torch.stack([snr, args.snr_gamma * torch.ones_like(timesteps)], dim=1).min(dim=1)[0] / snr
+                        )
+                    elif noise_scheduler.config.prediction_type == "v_prediction":
+                        mse_loss_weights = torch.stack([snr, args.snr_gamma * torch.ones_like(timesteps)], dim=1).min(
+                            dim=1
+                        )[0] / (snr + 1)
```
Comment (sayakpaul) [inline examples/kandinsky2_2/text_to_image/train_text_to_image_decoder.py:791]: `mse_loss_weights` doesn't need to be computed twice.

**Review by sayakpaul — COMMENTED** [review]
Thanks for your PR.

https://github.com/huggingface/diffusers/pull/6307/files#r1436118465 is my primary question.

File: examples/controlnet/train_controlnet_flax.py (hunk 1 above)
Comment (thuliu-yt16) [reply examples/controlnet/train_controlnet_flax.py:917]: I guess it is good to add another 'elif' branch here to take care of x0-prediction and leave the 'else' branch unchanged to allow for other potential prediction types in the future.

File: examples/controlnet/train_controlnet_flax.py (hunk 2 above)
Comment (thuliu-yt16) [reply examples/controlnet/train_controlnet_flax.py:911]: Well, I did not get the point. When `snr_gamma` is not specified, there is no need to compute `snr_loss_weights`.

Comment (sayakpaul) [reply examples/controlnet/train_controlnet_flax.py:911]: I am saying we always compute `snr_loss_weights` when `args.snr_gamma` is not None. Then we reuse it and make changes to it depending on the `noise_scheduler.config.prediction_type`.

Comment (thuliu-yt16) [reply examples/controlnet/train_controlnet_flax.py:911]: Get it. Do you mean exactly that we first create the weights as the default value min(SNR, gamma), and then we can divide it by SNR or SNR+1 respectively when we use eps/v-prediction?
I would say that both ways do not make much difference. Both are good to me.

Comment (sayakpaul) [reply examples/controlnet/train_controlnet_flax.py:911]: More concretely, I mean this:

```python
if args.snr_gamma is not None:
   snr = compute_snr(noise_scheduler, timesteps)
   mse_loss_weights = torch.stack([snr, args.snr_gamma * torch.ones_like(timesteps)], dim=1).min(dim=1)[0]
   
   if noise_scheduler.config.prediction_type == "v_prediction":
        mse_loss_weights = mse_loss_weights / (snr + 1)
```

Does that make sense?

Comment (thuliu-yt16) [reply examples/controlnet/train_controlnet_flax.py:911]: Of course. That's what I think exactly.

**Comment by thuliu-yt16** [comment]
done

**Comment by sayakpaul** [comment]
Will merge after the CI is green.

**Comment by siegelaaron94** [comment]
https://github.com/huggingface/diffusers/tree/main/examples/dreambooth was missed with this change

**Comment by sayakpaul** [comment]
@siegelaaron94 would you like to create a PR?
